metrics-api/src/main/java/software/amazon/swage/metrics/MetricContext.java

```diff
@@ -5,19 +5,23 @@
 import software.amazon.swage.collection.TypedMap;
 
 /**
- * The context in which a measurement is taken. A {MetricContext} contains the parameters of
+ * A {MetricContext} and its parent hierarchy contain the attributes of
  * the environment in which a measurement is taken.
  * In the physical world, these might define the effective time and place;
  * in a computer system, they might define the host, request, function, task or
  * other application-specific parameters.
  * <p/>
+ * The attributes in a {MetricContext} are the attributes which differentiate it from its parent.
+ * The attributes in the context can be combined with the attributes in its parent to
```

Comment:
What is the expected method to get all the attributes for a given context?  Not clear here whether it is in one context object or whether you now have to visit all the parents to get the information.  This seems to imply those looking at a context now need to fully traverse the ancestry.

Looking at the changed recorder code, it seems you actually do need to now look at the entire hierarchy.  Does every recorder now have to be updated to match this new approach?  It is a material change in the interface for any context that was created as a child, all code that wants to look at a context now needs to be aware there may be parents...

Reply:
Yes this is a material change in the interface. It looks to codify the hierarchical nature of metric contexts. Recorders will need to handle the hierarchy either in their reporting code or in their RecorderContext implementation.